src/p2p.rs

```diff
@@ -93,19 +93,31 @@ impl Connection {
             .collect();
         debug!("loading {} blocks", blockhashes.len());
         self.send(NetworkMessage::GetData(inv))?;
-        for hash in blockhashes {
-            match self
-                .recv()
-                .with_context(|| format!("failed to get block {}", hash))?
-            {
-                NetworkMessage::Block(block) => {
-                    assert_eq!(block.block_hash(), hash, "got unexpected block");
+
+        rayon::scope(|s| {
+            let (tx, rx) = crossbeam_channel::bounded(1);
+            s.spawn(|_| {
+                for (hash, block) in rx {
                     func(hash, block);
                 }
-                msg => bail!("unexpected {:?}", msg),
-            };
-        }
-        Ok(())
+            });
+
+            for hash in blockhashes {
+                match self
+                    .recv()
+                    .with_context(|| format!("failed to get block {}", hash))?
+                {
+                    NetworkMessage::Block(block) => {
+                        ensure!(block.block_hash() == hash, "got unexpected block");
+                        tx.send((hash, block))
+                            .context("disconnected from block processor")?;
+                    }
+                    msg => bail!("unexpected {:?}", msg),
+                };
+            }
+            drop(tx);
```

Comment:
I am quite surprised that you have to use drop method here, I heard you rarely have to call it thanks to ownership rules x')

Reply:
It's usually used as a comment for the reader if the author considers drop special in some way. In this case that would be "The channel gets closed here which causes stopping the other thread (iterator will return `None`)". It'd be equally valid without it, just less obvious.